File: practices/webflux/src/main/java/com/grizz/wooman/webflux/filter/SecurityWebFilter.java
package com.grizz.wooman.webflux.filter;

import com.grizz.wooman.webflux.auth.IamAuthentication;
import com.grizz.wooman.webflux.service.AuthService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.server.reactive.ServerHttpResponse;
import org.springframework.security.core.context.ReactiveSecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ServerWebExchange;
import org.springframework.web.server.WebFilter;
import org.springframework.web.server.WebFilterChain;
import reactor.core.publisher.Mono;
import reactor.util.context.Context;

@RequiredArgsConstructor
@Component
public class SecurityWebFilter implements WebFilter  {
    private final AuthService authService;

    @Override
    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
        final ServerHttpResponse resp = exchange.getResponse();
        String iam = exchange.getRequest().getHeaders()
                .getFirst("X-I-AM");

        if (iam == null) {
            resp.setStatusCode(HttpStatus.UNAUTHORIZED);
            return resp.setComplete();
        }

        return authService.getNameByToken(iam)
                .map(IamAuthentication::new)
                .flatMap(authentication -> {
                    return chain.filter(exchange)
                            .contextWrite(context -> {
                                Context newContext = ReactiveSecurityContextHolder
                                        .withAuthentication(authentication);

                                return context.putAll(newContext);
                            });
                })
                .switchIfEmpty(Mono.defer(() -> {
                    resp.setStatusCode(HttpStatus.UNAUTHORIZED);
                    return resp.setComplete();
                }));

    }
}

File: practices/coroutine-r2dbc-webflux/src/main/kotlin/com/grizz/wooman/coroutine/filter/SecurityWebFilter.kt
package com.grizz.wooman.coroutine.filter

import com.grizz.wooman.webflux.auth.IamAuthentication
import com.grizz.wooman.webflux.service.AuthService
import kotlinx.coroutines.reactor.awaitSingleOrNull
import kotlinx.coroutines.reactor.mono
import org.springframework.http.HttpStatus
import org.springframework.security.core.context.ReactiveSecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebFilter
import org.springframework.web.server.WebFilterChain
import reactor.core.publisher.Mono
import reactor.util.context.Context

@Component
class SecurityWebFilter(
    private val authService: AuthService
) : WebFilter {
    override fun filter(exchange: ServerWebExchange, chain: WebFilterChain): Mono<Void> {
        return mono {
            val resp = exchange.response
            val iam = exchange.request.headers
                .getFirst("X-I-AM")
            if (exchange.request.uri.path == "/api/users/signup") {
                return@mono chain.filter(exchange).awaitSingleOrNull()
            }
            if (iam == null) {
                resp.setStatusCode(HttpStatus.UNAUTHORIZED)
                return@mono resp.setComplete().awaitSingleOrNull()
            }

            val name = authService.getNameByToken(iam)
                .awaitSingleOrNull()
            if (name == null) {
                resp.setStatusCode(HttpStatus.UNAUTHORIZED)
                return@mono resp.setComplete().awaitSingleOrNull()
            }

            val authentication = IamAuthentication(name)

            chain.filter(exchange)
                .contextWrite { context: Context ->
                    val newContext = ReactiveSecurityContextHolder
                        .withAuthentication(authentication)
                    context.putAll(newContext)
                }.awaitSingleOrNull()
        }
    }
}
